Parameterize Graph by NodeType

AnishShah opened the pull request:
Is this Okay?

szymonm commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala:
```diff
@@ -24,12 +24,12 @@ import scala.util.Random
  * A Traverser traverses the graph in a certain order of nodes.
  */
-trait Traverser extends Iterator[Node] { self =>
+trait Traverser[NodeType <: Node] extends Iterator[Node] { self =>
```
why not Iterator[NodeType]?

szymonm replied:
also
```
trait Traverser[+NodeType <: Node]
```

szymonm commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/Graph.scala:
```diff
@@ -17,12 +17,12 @@ package com.twitter.cassovary.graph
  * and then traverse the graph using methods on that {@code Node}.
  */
-trait Graph {
+trait Graph[NodeType <: Node] {
```
probably
```
trait Graph[+NodeType <: Node]
```
is what we want

pankajgupta replied:
prefer a shorter V instead of 'NodeType', so
trait Graph[+V <: Node]

szymonm commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala:
```diff
@@ -396,7 +396,7 @@ trait QueueBasedTraverser extends Traverser {
  * A traverser that keeps track of first depth of visiting a given node.
  */
-trait DepthTracker extends QueueBasedTraverser {
+trait DepthTracker[NodeType <: Node] extends QueueBasedTraverser[NodeType] {
```
I wondering if we cannot use just
```
trait DepthTracker extends QueueBasedTraverser[Node]
```
as DepthTracker itself does not use information about particular Node type.

pankajgupta replied:
@szymonm while it technically does not have to be as you point out, there is a case for making it because the trait DepthTracker "is a" QueueBasedTraverser (because of 'extends' keyword) and it is a bit odd to not type parametrize it. e.g., here do we just want to have a QueueBasedTraverser of Node, and not of more specific NodeTypes? If so (and I am fine with it), your suggestion makes sense.

szymonm replied:
QueueBasedTraverser should definitely be of NodeType.

pankajgupta replied:
but DepthTracker only needs QueueBasedTraverser[Node] ? If so, okay to not
have type parametrization in DepthTracker.

On Wed, Feb 4, 2015 at 8:39 AM, Szymon <notifications@github.com> wrote:

> In
> cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala
> <https://github.com/twitter/cassovary/pull/148#discussion_r24097040>:
>
> > @@ -396,7 +396,7 @@ trait QueueBasedTraverser extends Traverser {
> >  /**
> >   * A traverser that keeps track of first depth of visiting a given node.
> >   */
> > -trait DepthTracker extends QueueBasedTraverser {
> > +trait DepthTracker[NodeType <: Node] extends QueueBasedTraverser[NodeType] {
>
> QueueBasedTraverser should definitely be of NodeType.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/cassovary/pull/148/files#r24097040>.
>

AnishShah replied:
so, just `QueueBasedTraverser[Node]` ? @szymonm ?

pankajgupta replied:
Let's go with that for now.

szymonm replied:
Thought about this. I think 
```
trait DepthTracker  extends QueueBasedTraverser[Node]
```
is ok. Thanks to covariance QueueBasedTraverser[NodeType] <: QueueBasedTraverser[Node] so we can extend it with DepthTracker.

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/DirectedGraph.scala:
```diff
@@ -33,7 +33,7 @@ object StoredGraphDir extends Enumeration {
  * and then traverse the graph using methods on that {@code Node}.
  */
-trait DirectedGraph extends Graph with Iterable[Node] {
+trait DirectedGraph extends Graph[Node] with Iterable[Node] {
```
DirectedGraph should also be of type [+NodeType <: Node]

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala:
```diff
@@ -71,10 +71,10 @@ trait BoundedIterator[T] extends Iterator[T] {
  * @param filterHomeNodeByNumEdges filter home node by number of edges
  */
-class RandomTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
+class RandomTraverser[NodeType <: Node](graph: Graph[NodeType], dir: GraphDir, homeNodeIds: Seq[Int],
```
use +NodeType instead of NodeType everywhere.

AnishShah commented:
Please have a look. I'm not sure about `ArrayBasedDirectedGraph`, `SharedArrayBasedDirectedGraph` and also Benchmark files. But, I have checked everything and it works fine.

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/PageRank.scala:
```diff
@@ -42,7 +42,7 @@ object PageRank {
    * @return An array of doubles, with indices corresponding to node ids
    */
-  def apply(graph: DirectedGraph, params: PageRankParams): Array[Double] = {
+  def apply[V <: Node](graph: DirectedGraph[V], params: PageRankParams): Array[Double] = {
```
No need to parametrize in this file. It uses only node.id that you can get by just Directedgraph[Node]

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/TriangleCount.scala:
```diff
@@ -28,7 +27,7 @@ case class TriangleCountParameters(edgeReservoirSize : Int, wedgeReservoirSize :
  */
 object TriangleCount {
-  def apply(graph: DirectedGraph, parameters: TriangleCountParameters): (Double, Double) = {
+  def apply[V <: Node](graph: DirectedGraph[V], parameters: TriangleCountParameters): (Double, Double) = {
```
Ditto, no need to parametrize.

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/AbstractCentrality.scala:
```diff
@@ -21,7 +21,7 @@ import com.twitter.cassovary.graph.{Node, DirectedGraph}
  * {@note The centrality values are stored in an internal array indexed by node's id. Hence, do not use for graphs with very large ids.}
  */
-abstract class AbstractCentrality(graph: DirectedGraph) extends Centrality {
+abstract class AbstractCentrality[+V <: Node](graph: DirectedGraph[V]) extends Centrality {
```
Ditto, no need to parametrize.

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/SharedArrayBasedDirectedGraph.scala:
```diff
@@ -255,7 +255,7 @@ class SharedArrayBasedDirectedGraph private (nodeIdSet: Array[Byte], offsetTable
     reverseDirEdgeArray: Option[Array[Array[Int]]], maxId: Int, val nodeWithOutEdgesMaxId: Int,
     val nodeWithOutEdgesCount: Int, val nodeCount: Int, val edgeCount: Long,
-    val storedGraphDir: StoredGraphDir) extends DirectedGraph {
+    val storedGraphDir: StoredGraphDir) extends DirectedGraph[Node] {
```
for now, this is okay, but we will change it in a separate PR/issue.

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala:
```diff
@@ -24,21 +24,21 @@ import scala.util.Random
    * we know that node must exist, safe to assume graph.getNodeById return non-None value
    */
-  protected def getExistingNodeById(graph: Graph, id: Int) = graph.getNodeById(id).get
+  protected def getExistingNodeById[V <: Node](graph: Graph[V], id: Int) = graph.getNodeById(id).get
```
By defining a new [V <: Node] here you have shadowed the V in Traverser[V]

what you actually want here is:
getExistingNodeById[W >: V](graph: Graph[W]...)


see http://docs.scala-lang.org/tutorials/tour/variances.html

AnishShah replied:
Not sure if it should be `W >: V` . It means, W should be either `V` or its super type. But, here V is only the sub types of Node, which breaks the type bound, I guess.

AnishShah replied:
It should be `W <: Node`, I guess.

pankajgupta replied:
No I think W >: V is right. A traverser of type V can do something with
graph of type V or supertype but can not of subtypes.

On Saturday, February 7, 2015, Anish Shah <notifications@github.com> wrote:

> In
> cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala
> <https://github.com/twitter/cassovary/pull/148#discussion_r24290377>:
>
> >    /**
> >     * Traverser only visits node Ids listed in some node's edges, and thus
> >     * we know that node must exist, safe to assume graph.getNodeById return non-None value
> >     */
> > -  protected def getExistingNodeById(graph: Graph, id: Int) = graph.getNodeById(id).get
> > +  protected def getExistingNodeById[V <: Node](graph: Graph[V], id: Int) = graph.getNodeById(id).get
>
> It should be W <: Node, I guess.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/cassovary/pull/148/files#r24290377>.
>

AnishShah replied:
I think, it is breaking the type bound.
I'm getting this error `type arguments [W] do not conform to trait Graph's type parameter bounds [+V <: com.twitter.cassovary.graph.Node]`

Here is some rough code
```
scala> trait Node
defined trait Node

scala> trait Graph[+V <: Node]
defined trait Graph

scala> trait Traverser[+V <: Node] {
     |   protected def someFunction[W >: V](graph: Graph[W]) = println("hi")
     | }
<console>:10: error: type arguments [W] do not conform to trait Graph's type parameter bounds [+V <: Node]
         protected def someFunction[W >: V](graph: Graph[W]) = println("hi")
 
```

pankajgupta replied:
try [W >: V <: Node]

AnishShah replied:
Yeah! That was working. Cool :)
On 15 Feb 2015 13:44, "Pankaj Gupta" <notifications@github.com> wrote:

> In
> cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala
> <https://github.com/twitter/cassovary/pull/148#discussion_r24720640>:
>
> >    /**
> >     * Traverser only visits node Ids listed in some node's edges, and thus
> >     * we know that node must exist, safe to assume graph.getNodeById return non-None value
> >     */
> > -  protected def getExistingNodeById(graph: Graph, id: Int) = graph.getNodeById(id).get
> > +  protected def getExistingNodeById[V <: Node](graph: Graph[V], id: Int) = graph.getNodeById(id).get
>
> try [W >: V <: Node]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/cassovary/pull/148/files#r24720640>.
>

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala:
```diff
@@ -363,7 +363,7 @@ trait QueueBasedTraverser extends Traverser {
    * Performs action needed when visiting a node.
    */
-  protected def processNode(node : Node): Node = {
+  protected def processNode[V <: Node](node : V): V = {
```
Same problem as above U/W discussion. See if you can simply have processNode(Node) here. Also, seems like the return value is not being used so far, so return Unit.

AnishShah replied:
return value is used in `next()`

AnishShah replied:
Also, can't keep `processNode(Node)` since return type of `next()` should be the subtype of `Node` .

Should I also parameterize `visitNode()` ?

pankajgupta replied:
In next(), instead of 
```processNode(getExistingNodeById(graph, nextId))```
perhaps do this:
```
...
val node = getExistingNodeById(graph, nextId)
processNode(node)
node
```

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/bipartite/IterativeLinkAnalyzer.scala:
```diff
@@ -33,7 +33,7 @@ case class SuppliedNodeInfo(val node: Node, val initialIterationWeight: Double)
  * @param numTopContributors The number of contributors (reasons) to be kept for each score
  */
-class IterativeLinkAnalyzer(graphUtils: GraphUtils, resetProbOnLeft: Double,
+class IterativeLinkAnalyzer[+V <: Node](graphUtils: GraphUtils[V], resetProbOnLeft: Double,
```
leave as GraphUtils[Node]

pankajgupta commented on cassovary-core/src/test/scala/com/twitter/cassovary/graph/GraphBehaviours.scala:
```diff
@@ -19,7 +19,7 @@ import org.scalatest.{Matchers, WordSpec}
 trait GraphBehaviours extends Matchers {
```
just make this trait of type [V <: Node] and simply use DirectedGraph[V] in all methods below without specifying [V<; Node] in every method

pankajgupta commented on cassovary-core/src/test/scala/com/twitter/cassovary/graph/TraverserSpec.scala:
```diff
@@ -248,7 +248,7 @@ class TraverserSpec extends WordSpec with MockitoSugar with Matchers {
       val dir = GraphDir.InDir
       val trav = new DepthFirstTraverser(graph, dir, Seq(10), Walk.Limits(None, None, Some(10L)))
-        with PathLengthTracker with DiscoveryAndFinishTimeTracker
+        with PathLengthTracker[Node] with DiscoveryAndFinishTimeTracker[Node]
```
will be nice to add a test on any traverser on any subtype of Node to test the new parametrization.

AnishShah replied:
OK. Will do it.

AnishShah replied:
I need help with writing tests. Should I create `TestBipartiteGraph.scala` for `Bipartite Node` and `TestDynamicGraph.scala`  for `Dynamic Node` and then use it in `TraverserSpec.scala` ?

pankajgupta replied:
Just add a new bipartite graph to TestGraph.scala and use it in TraverserSpec. We are looking for one simple test to test the new functionality, and do not need all the methods / classes of Traverser tested again

pankajgupta commented:
We do need to move all DirectedGraph implementations to be parametrized by node type as well. But lets do it in a separate PR.

pankajgupta commented:
LGTM otherwise

AnishShah commented:
I'm busy this week. I won't be able to work on this until Friday. Sorry.

AnishShah commented:
I have added the tests

1. I refactored the code and moved `BipartiteGraphSingleSide & DoubleSide` to TestGraphs and I'm using that in `BipartiteGraphSpec.scala` and `TraverserSpec.scala`. Is it OK?

2. While writing the test, I found out a bug. in [this line](https://github.com/AnishShah/cassovary/commit/776267fcac7645e4d3f0dc4fc1308f638771de24#diff-e7998641f7338821ff1199baf366994aL331) , `depth(node.id).get` fails for the first node since depth is not assigned to the first node. So, I changed it to getOrElse(-1). Am I correct? 

Please tell me if there are any other changes needed.

pankajgupta commented:
1. OK

2. Curious how you ran into the bug. there is   `enqueue(homeNodeIds, None)` to make sure homeNodeIds are assigned depth in DepthTracker.

AnishShah commented:
@pankajgupta So sorry, I misinterpreted the bug. It is actually because for Bipartite Graph LeftNode,
`enqueue(homeNodeIds, None)` saves depth for negative number and in `chooseNodesToEnqueue` it asks for depth of LeftNode but with positive number. 

I will need your help on this. Should I save negative id for LeftNode? What do you suggest?

pankajgupta commented:
No worries.
I think that might be a bug in BipartiteGraph (possibly where class LeftNode in BipartiteGraph.scala should have val id = -nodeId). I will suggest removing the Traverser tests for BipartiteGraph, closing this PR and tackling the traverser tests for BipartiteGraph by fixing this bug in a separate PR.

AnishShah commented:
Cool
On 18 Feb 2015 12:00, "Pankaj Gupta" <notifications@github.com> wrote:

> No worries.
> I think that might be a bug in BipartiteGraph (possibly where class
> LeftNode in BipartiteGraph.scala should have val id = -nodeId). I will
> suggest removing the Traverser tests for BipartiteGraph, closing this PR
> and tackling the traverser tests for BipartiteGraph by fixing this bug in a
> separate PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/cassovary/pull/148#issuecomment-74819587>.
>

pankajgupta commented:
Sorry I wasn't clear. Let's finish this PR first and merge to master without the BipartiteGraph traverser tests. Then add BipartiteGraph traverser tests in a separate PR.

AnishShah commented:
I have remove the tests, changed `getOrElse(-1)` to `get` but I have still kept the bipartiteGraph in TestGraphs.

Please tell me what changes are needed now?

pankajgupta commented:
@AnishShah please follow up with (1) the bipartite graph tests we removed from this PR, and (2) Wmove all DirectedGraph implementations to be parametrized by node type as well. Please make new issues for each one if you don't mind.
